sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/GroupIntoBatches.java

```diff
@@ -100,8 +99,7 @@ public long getBatchSize() {
         ParDo.of(new GroupIntoBatchesDoFn<>(batchSize, allowedLateness, keyCoder, valueCoder)));
   }
 
-  @VisibleForTesting
-  static class GroupIntoBatchesDoFn<K, InputT>
+  public static class GroupIntoBatchesDoFn<K, InputT>
```

Comment:
I don't think we want to make `GroupInToBatchesDoFn` as `public`. Instead of adding property in `translateFn `, we can do it at transform level, where you can check `transform instanceof GroupIntoBatches`, for example: https://github.com/apache/beam/blob/master/runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowPipelineTranslator.java#L752

Reply:
Yeah I did try matching the transform by `transform instanceof GroupIntoBatches` but got error:
```
Inconvertible types; cannot cast 'org.apache.beam.runners.dataflow.PrimitiveParDoSingleFactory.ParDoSingle<InputT,OutputT>' to 'org.apache.beam.sdk.transforms.GroupIntoBatches'
```
Any ideas?

Reply:
Have you tried `registerTransformTranslator `?

Reply:
I assume that we would still need to access to `GroupIntoBatchesDoFn` if we register a separate translator, to fill USER_FN for example: https://github.com/apache/beam/blob/master/runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowPipelineTranslator.java#L1241.

Can the same transform go though multiple translators? Is that what you were suggesting?

Reply:
I was thinking about we could check `GroupIntoBatches` directly but it turns out that the translation happens after transform expansions, where we only can get `ParDo(GroupIntoBatchesDoFn)`. One thing we can do to avoid exposing `GroupIntoBatchesDoFn` is to make `DoFnSignature` understand that this is a `shardable stateful DoFn`, like what we do for `RESTRICTION_ENCODING`.  @lukecwik Do you have any suggestion here?

Reply:
There have been a few ways this has been done in the past:
* (easiest), record which transforms need this property within the DataflowRunner and then lookup this information during translation (e.g. [doesPCollectionRequireIndexedFormat](https://github.com/apache/beam/blob/b1849ed09fb236906ff0b83b0f394c08b05d4b3c/runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowRunner.java#L1260))
* replace the public GroupIntoBatches transform with a Dataflow specific `primitive` that makes any additional information visible that is needed during translation (e.g. [DataflowRunner.CombineGroupedValues](https://github.com/apache/beam/blob/b1849ed09fb236906ff0b83b0f394c08b05d4b3c/runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowPipelineTranslator.java#L799))

Reply:
Thanks! I applied the first solution recommended above and updated the PR accordingly - I added an override for streaming GroupIntoBatches which records the input PCollection, and also a transform matcher to pass the override validation. PTAL.